[FEATURE] Define `rg.Settings` based on `datasets.Features`

burtenshaw opened the pull request:
# Description
<!-- Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

This PR define a module that interacts with the datasets server API to create a Settings object based on  datasets features.

It can be used like this:

## Via `rg.Settings`

```python
from argilla import Settings, Dataset
from datasets import load_dataset

settings = Settings.from_hub(repo_id= "yahma/alpaca-cleaned")
dataset = Dataset(name= "alpaca-cleaned", settings=settings)
dataset.create()
hf_dataset = load_dataset(path=repo_id, split="train[:10]")
dataset.records.log(hf_dataset)
```

## Via `rg.Dataset`

```python
from argilla import Dataset

dataset = Dataset.from_hub(repo_id="yahma/alpaca-cleaned")
```

from_hub has 3 uses:
- used with settings
- used on an argilla config
- used with nothing

## Things to note

- It's implemented as a helper module with functions, in case we want to migrate it to the server.
- It's pretty cautious in it's error handling from datasets-server, we might want to add a fall back, and more warnings.
- I think we should upgrade integration testing when we use this in the `from_hub` method.

## Next Steps
The next step is to integrate it with `from_hub`, so this works:

```python
dataset = rg.Dataset.from_hub(repo_id="yahma/alpaca-cleaned")
```



**Type of change**
<!--  Please delete options that are not relevant. Remember to title the PR according to the type of change  -->

- New feature (non-breaking change which adds functionality)

**How Has This Been Tested**
<!--  Please add some reference about how your feature has been tested.  -->

**Checklist**
<!--  Please go over the list and make sure you've taken everything into account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

burtenshaw committed:
test: add basic unit tests for feature reponses

burtenshaw committed:
feat: implement feature for settings definition in _helpers

burtenshaw committed:
feat: integrate datasets server helper with settings resource

burtenshaw committed:
fic: resolve imports in helper

davidberenstein1957 reviewed (commented):
@burtenshaw, can you update the PR and only include the relevant commits?

frascuchon commented on argilla/src/argilla/_helpers/_datasets_server.py:
Why not to move this under the _io/_datasets/ module?

burtenshaw replied:
I mainly put it there because of the format, but `datasets._io` works. But  `settings._io` could make the most sense.

burtenshaw replied:
Moved it to `settings._io`

frascuchon commented on argilla/src/argilla/_helpers/_datasets_server.py:
```diff
@@ -0,0 +1,175 @@
+    url = f"{DATASETS_SERVER_BASE_URL}/info?dataset={repo_id}"
+
+    try:
+        response = requests.get(url, headers=DATASETS_SERVER_HEADERS, timeout=DATASETS_SERVER_TIMEOUT)
```
Since we're already using httpx for http request, maybe we could align this and use the same package here.

frascuchon commented on argilla/src/argilla/_helpers/_datasets_server.py:
```diff
@@ -0,0 +1,175 @@
+    try:
+        dataset_info = response_json["dataset_info"]
+    except KeyError as e:
+        raise DatasetsServerException() from e
```
I think we can wrap all the caught exceptions in one `except` block:
```suggestion
    try:
        response = requests.get(url, headers=DATASETS_SERVER_HEADERS, timeout=DATASETS_SERVER_TIMEOUT)
        response.raise_for_status()    
        response_json = response.json()
        dataset_info = response_json["dataset_info"]    
    except (requests.exceptions.RequestException, requests.exceptions.HTTPError, KeyError, Exception) as e:
        raise DatasetsServerException() from e
```

burtenshaw committed:
rename: move datasets._export to datasets._io

frascuchon commented on argilla/src/argilla/_helpers/_datasets_server.py:
```diff
@@ -0,0 +1,175 @@
+    available_configs = list(dataset_info.keys())
+
+    if len(available_configs) > 1 and config is None:
+        raise ValueError("Multiple configurations found. Please specify a config.")
```
I think is easier for users just warn and read the first available config. What do you think?

burtenshaw replied:
Some datasets require a specific config to be used, but they're rare. Let's go with this for now.

frascuchon commented on argilla/src/argilla/_helpers/_datasets_server.py:
```diff
@@ -0,0 +1,175 @@
+    elif len(available_configs) == 1:
+        config = available_configs[0]
+    elif config not in available_configs:
+        raise ValueError(f"Configuration '{config}' not found.")
```
In which scenario this is possible? `config` is a required parameter for getting rows.

burtenshaw replied:
If the user defines a non-existent config.

frascuchon commented on argilla/src/argilla/_helpers/_datasets_server.py:
```diff
@@ -0,0 +1,175 @@
+
+        elif feature_type == FeatureType.TEXT:
+            fields.append(TextField(name=f"{name}_field"))
+            questions.append(TextQuestion(name=f"{name}_question"))
```
Do we want to add a required text question for each text feature found?

burtenshaw replied:
Ideally not, but because datasets require at least one required question we need to do something here. Any suggestions?

frascuchon replied:
What about adding a default question if questions are not mapped at the end of the process? We need to create a dataset without the required questions, but we need to evaluate the implications of doing this.

burtenshaw replied:
That would get us moving. We could go for something like a TextQuestion named 'comment'?

frascuchon commented on argilla/src/argilla/_helpers/_datasets_server.py:
```diff
@@ -0,0 +1,175 @@
+        elif feature_type == FeatureType.LABEL:
+            fields.append(TextField(name=f"{name}_field"))
+            questions.append(
+                LabelQuestion(
```
What about MultiLabelQuestion ? Do we have a mechanism to identify Multi vs Label questions? If not maybe is more flexible to create MultiLabelQuestion

burtenshaw replied:
There is a separate representation for multilabel features in datasets, so we know that this is a `LabelQuestion` compatible field.

frascuchon commented on argilla/src/argilla/_helpers/_datasets_server.py:
```diff
@@ -0,0 +1,175 @@
+            mapping[name] = f"{name}_field"
+
+        elif feature_type == FeatureType.LABEL:
+            fields.append(TextField(name=f"{name}_field"))
```
Do we want to create a text field including label values? I was thinking of passing them as suggestionsvalues

burtenshaw replied:
Because of the mapping, they would appear as suggestion like this. I can remove this if you think it's overkill.

burtenshaw committed:
rename: move from datasets._io to settings._io

frascuchon commented on argilla/src/argilla/_helpers/_datasets_server.py:
```diff
@@ -0,0 +1,175 @@
+            pass
+
+        elif feature_type == FeatureType.TEXT:
+            fields.append(TextField(name=f"{name}_field"))
```
Why we cannot use the name directly?
```python
fields.append(TextField(name=name))
```

In general, do we need to apply some name conversion to adapt feature names to argilla setting names? (lower case,...)

burtenshaw replied:
We discussed using values as both field and question so that it's as flexible as possible. Based on that, I need to differentiate the names for attributes and then map them.

burtenshaw committed:
test: updated tests for move

frascuchon commented on argilla/src/argilla/settings/_resource.py:
```diff
@@ -257,6 +258,13 @@ def from_json(cls, path: Union[Path, str]) -> "Settings":
+    def from_hub(cls, repo_id: str, version: str = "latest") -> "Settings":
+        """Load the settings from the Hub"""
+
+        settings = build_settings_from_repo_id(repo_id)
```
We're not using the `config` parameter here. I would say to remove and fetch the first configuration when multiple are found.

frascuchon commented on argilla/src/argilla/_helpers/_datasets_server.py:
```diff
@@ -0,0 +1,175 @@
+    FLOAT = "float"
+
+
+def _get_dataset_features(repo_id: str, config: Optional[str] = None) -> Dict[str, Dict[str, Any]]:
```
Regarding this comment https://github.com/argilla-io/argilla/pull/5472/files#r1751978426
```suggestion
def _get_dataset_features(repo_id: str) -> Dict[str, Dict[str, Any]]:
```

frascuchon commented on argilla/src/argilla/_helpers/_datasets_server.py:
```diff
@@ -0,0 +1,175 @@
+    )
+
+
+def _define_settings_from_features(features: Union[List[Dict], Dict[str, Any]], config=None) -> "Settings":
```
Regarding this comment https://github.com/argilla-io/argilla/pull/5472/files#r1751978426
```suggestion
def _define_settings_from_features(features: Union[List[Dict], Dict[str, Any]]) -> "Settings":
```

frascuchon replied:
I think the `config` argument is not used on this function. We can remove it.

burtenshaw committed:
feat: reimplement datasets server module in httpx

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
feat: simplify error handling

burtenshaw committed:
Merge branch 'feature/features-based-settings' of https://github.com/argilla-io/argilla into feature/features-based-settings

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
feat: always use first config for a dataset

burtenshaw committed:
Merge branch 'feature/features-based-settings' of https://github.com/argilla-io/argilla into feature/features-based-settings

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon commented on argilla/src/argilla/settings/_io/_hub.py:
```diff
@@ -0,0 +1,167 @@
+        else:
+            warnings.warn(f"Feature '{name}' has an unsupported type. Skipping. Feature type: {feature_type}")
+
+    settings = Settings(fields=fields, questions=questions, metadata=metadata, mapping=mapping)
```
Maybe we can add this if no questions were mapped (with a warning). So we can remove the TextQuestion for text features. What do you think?

```suggestion
    if not questions:
        questions.append(rg.TextQuestion(name="comment",...)
    settings = Settings(fields=fields, questions=questions, metadata=metadata, mapping=mapping)
```

burtenshaw replied:
Makes sense

frascuchon commented on argilla/src/argilla/settings/_io/_hub.py:
```diff
@@ -0,0 +1,167 @@
+    return settings
+
+
+def build_settings_from_repo_id(repo_id: str):
```
>  dealing with  datasets columns as fields, metadata, and questions. When should we do this? When is it overkill?

Maybe as a first flexible step, allow passing an optional features mapping dict could help to customize and evolve this? For example for this [dataset](https://huggingface.co/datasets/AI-MO/NuminaMath-CoT?row=0)

```python

features_mapping = { "source": "metadata",  "solution": "question" }
build_settings_from_repo_id("AI-MO/NuminaMath-CoT", features_mapping)
```

burtenshaw committed:
[FEATURE] encode class labels as names not ints (#5474)

# Description
<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->

This PR add to the settings from_hub PR by supporting ClassLabels coming
from datasets package.

It renders a label question with names based on the ClassLabel.

It uncasts the 'label.suggestion' to a corresponding class label.

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- New feature (non-breaking change which adds functionality)

burtenshaw committed:
feat: add default comment question if questions are missing

Co-authored-by: Paco Aranda <frascuchon@gmail.com>

burtenshaw committed:
feat: support mapping for columns to argilla dataset attributes

burtenshaw committed:
fix: IntegerMetadataProperty

burtenshaw committed:
feat: add feature mapping to settings resource

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
fix: define literal in resource

burtenshaw committed:
fox: make feature mapping default to NOne

burtenshaw committed:
fix: catch metadata when exporting to datasets

burtenshaw committed:
Merge branch 'develop' into feature/features-based-settings

burtenshaw committed:
fix: validate that image records can be visualised in readme

burtenshaw committed:
fix: deal with persisting images that don't exist as strings

burtenshaw committed:
test: make feature import naming consistent

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
fix: remove config variable

burtenshaw committed:
fix: raise error for datasets without fields

burtenshaw committed:
test: update test for reduce fields and questions and mapping

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
feat: resolve naming in record io dataset >> hf_dataset private to public

burtenshaw committed:
fix: resolve dataset casting in hub io

burtenshaw committed:
Merge branch 'feature/features-based-settings' of https://github.com/argilla-io/argilla into feature/features-based-settings

frascuchon commented on argilla/src/argilla/settings/_io/_hub.py:
```diff
@@ -0,0 +1,203 @@
+    feature_mapping = feature_mapping or {}
+    mapped_questions = [key for key, value in feature_mapping.items() if value == "question"]
+    mapped_fields = [key for key, value in feature_mapping.items() if value == "field"]
+    mapped_metadata = [key for key, value in feature_mapping.items() if value == "metadata"]
```
Do we need these extra lists? Maybe we can access the feature_mapping dictionary directly:

```python
   defined_mapping = feature_mapping.get(name)
   if defined_mapping == "question" ...
```
with some default values depending on the extracted feature type? (For example, field for text features, and question for label features?

frascuchon commented on argilla/src/argilla/settings/_io/_hub.py:
```diff
@@ -0,0 +1,203 @@
+            pass
+
+        elif feature_type == FeatureType.TEXT:
+            if name not in mapped_questions:
```
Why not a positive check?
```python
   defined_mapping = feature_mapping.get(name, "field")
   if defined_mapping == "question":
       questions.append(...)
   elif defined_mapping == "field":
       fields.append(...)
   elif defined_mapping == "metadata": 
       # allow extra metadata = True?
```

burtenshaw replied:
You're right that the logic is a bit weird, but your suggestion changes the logic. The implemented logic does this:

```
defined_mapping == "field" == "field"
defined_mapping =="question" == "question"
defined_mapping == None == "question" and "field"

defined_mapping =="question" == "question"
defined_mapping =="metadata" == "metadata"
defined_mapping == None == "question" and "metadata"
```

We can change logic if you thing that's wrong as well.

burtenshaw replied:
In simple terms, the implementation says: if a feature is mapped, assign to one attribute, otherwise assign to both.

frascuchon commented on argilla/src/argilla/settings/_io/_hub.py:
```diff
@@ -0,0 +1,203 @@
+                mapping[name].append(name)
+            if name not in mapped_questions:
+                metadata.append(TermsMetadataProperty(name=f"{name}_metadata", options=names))
+                mapping[name].append(f"{name}_metadata")
```
same as the comment before:
```python
if feature_mappings.get(name) == "question":
    questions.append(...)
elif feature_mapping(name ) == "metadata":
    metadata.append(...)
```

frascuchon commented on argilla/src/argilla/settings/_io/_hub.py:
```diff
@@ -0,0 +1,203 @@
+
+def build_settings_from_repo_id(repo_id: str, feature_mapping: Optional[Dict[str, str]] = None) -> "Settings":
+    dataset_info = _get_dataset_features(repo_id)
+    return _define_settings_from_features(dataset_info, feature_mapping)
```
```suggestion
    dataset_features = _get_dataset_features(repo_id)
    return _define_settings_from_features(dataset_features, feature_mapping)
```

frascuchon commented on argilla/src/argilla/settings/_io/_hub.py:
```diff
@@ -0,0 +1,203 @@
+    mapping = {
+        key: value[0] if isinstance(value, list) and len(value) == 1 else tuple(value)
+        for key, value in feature_mapping.items()
+    }
```
I'm a bit confused with these lines. Can you share an example of the `feature_mapping` content?

burtenshaw replied:
This is a mistake, it should be on `mapping`:

```python
    mapping = {
        key: value[0] if isinstance(value, list) and len(value) == 1 else tuple(value)
        for key, value in mapping.items()
    }
```

frascuchon replied:
Okay, so this code is to normalize mapping values since we don't support tuples with one single value, right?

burtenshaw replied:
That's correct.

frascuchon commented on argilla/src/argilla/settings/_resource.py:
```diff
@@ -257,6 +258,15 @@ def from_json(cls, path: Union[Path, str]) -> "Settings":
 
+    @classmethod
+    def from_hub(
+        cls, repo_id: str, feature_mapping: Optional[Dict[str, Literal["question", "field", "metadata"]]] = None
```
Based on this feature_mapping definition, I think the code above will fail, since it expects a dict[str, str]

frascuchon commented on argilla/tests/unit/test_settings/test_settings_from_features.py:
```diff
@@ -0,0 +1,120 @@
+        "image_column": {"_type": "Image"},
+        "label_column": {"_type": "ClassLabel", "names": ["A", "B"]},
+    }
+    settings = _define_settings_from_features(features, feature_mapping={"text_column": "question"})
```
Does the resultant settings mapping match the expected one?

burtenshaw replied:
Only necessary Argilla dataset attributes are mapped. I.e. ones that get renamed to deal with multiple attributes for the same column.

frascuchon commented on argilla/tests/unit/test_settings/test_settings_from_features.py:
```diff
@@ -0,0 +1,120 @@
+    assert isinstance(settings.questions[0], rg.TextQuestion)
+    assert settings.questions[0].name == "text_column_question"
+    assert isinstance(settings.questions[1], rg.LabelQuestion)
+    assert settings.questions[1].name == "label_column"
```
Should be this the expected mapping?
```suggestion
    assert settings.questions[1].name == "label_column"
    assert settings.mapping == {  "text_column":  "text_columnt_question" }
```

davidberenstein1957 commented on argilla/src/argilla/datasets/_io/_hub.py:
```diff
@@ -249,3 +249,27 @@ def _get_dataset_split(hf_dataset: "HFDataset", split: Optional[str] = None, **k
+                    sample_huggingface_record[key] = value
+                except TypeError:
+                    try:
+                        sample_huggingface_record[key] = pil_to_data_uri(value)
```
can't we check if it is an image specifically?

davidberenstein1957 commented on argilla/src/argilla/records/_io/_datasets.py:
```diff
@@ -26,6 +27,125 @@
+        features = hf_dataset.features.copy()
+        features[column] = Value("string")
+        hf_dataset = hf_dataset.map(
+            lambda x: {column: hf_dataset.features[column].int2str(x[column])}, features=features
```
is `int2str` always defined? I thought that sometimes datasets don't contain the proper mapping.

burtenshaw replied:
If it's a class label it will have an int2str attribute.

davidberenstein1957 commented on argilla/src/argilla/records/_io/_datasets.py:
```diff
@@ -26,6 +27,125 @@
+    for column in columns:
+        features = hf_dataset.features.copy()
+        features[column] = Image()
+        casted_hf_dataset = hf_dataset.map(
+            function=lambda batch: {column: [uncast_image(sample) for sample in batch]},
+            with_indices=False,
+            batched=True,
+            input_columns=[column],
+            remove_columns=[column],
+            features=features,
+        )
+        try:
+            casted_hf_dataset[0]
```
is this correct? when a `None` return, 
```suggestion
        
        try:
            casted_hf_dataset = hf_dataset.map(
                function=lambda batch: {column: [uncast_image(sample) for sample in batch]},
                with_indices=False,
                batched=True,
                input_columns=[column],
                remove_columns=[column],
                features=features,
            )
            casted_hf_dataset[0]
```

burtenshaw replied:
`casted_hf_dataset[0]` will fail if the image value is not compatible with dataset serialization. Why should the map execution be inside the exception?

davidberenstein1957 commented on argilla/src/argilla/records/_io/_datasets.py:
```diff
@@ -64,20 +182,18 @@ def _record_dicts_from_datasets(dataset: HFDataset) -> List[Dict[str, Union[str,
         record_dicts = []
-        for example in dataset:
+        for example in hf_dataset:
             record_dicts.append(example)
```
```suggestion
        record_dicts.append([example for example in hf_dataset])
```

frascuchon replied:
or:
```python
record_dicts.extend(hf_dataset)
```
?

burtenshaw replied:
Dataset samples are serialised on iteration. I can use a comprehension if that prettier.

davidberenstein1957 commented on argilla/src/argilla/records/_io/_datasets.py:
```diff
@@ -98,64 +222,15 @@ def _get_image_features(dataset: "HFDataset") -> List[str]:
+            if isinstance(feature, Image):
+                casted_features[Image].append(name)
+            if isinstance(feature, ClassLabel):
+                casted_features[ClassLabel].append(name)
```
so, there is not differentiation between single/multi-label?

burtenshaw replied:
`ClassLabel` == `rg.LabelQuestion` multi label in datasets is a `Sequence[ClassLabel]` but it's out of scope.

davidberenstein1957 commented on argilla/src/argilla/settings/_io/_hub.py:
```diff
@@ -0,0 +1,203 @@
+DATASETS_SERVER_TIMEOUT = 30
+
+
+class FeatureType(Enum):
+    TEXT = "text"
+    CHAT = "chat"
+    IMAGE = "image"
+    LABEL = "label"
+    INT = "int"
+    FLOAT = "float"
```
same here, no differentiation between single and multi label?

burtenshaw replied:
Multilabel is out of scope.

davidberenstein1957 commented on argilla/src/argilla/settings/_io/_hub.py:
```diff
@@ -0,0 +1,203 @@
+
+DATASETS_SERVER_BASE_URL = "https://datasets-server.huggingface.co"
+DATASETS_SERVER_HEADERS = {"Accept": "application/json"}
+DATASETS_SERVER_TIMEOUT = 30
```
do we have back-off retry policy somewhere else by chance?

burtenshaw replied:
Not that I know of, for the datasets server API.

davidberenstein1957 commented on argilla/src/argilla/settings/_io/_hub.py:
```diff
@@ -0,0 +1,203 @@
+            dataset_info = response_json["dataset_info"]
+            available_configs = list(dataset_info.keys())
+
+            if len(available_configs) > 1:
+                warnings.warn("Multiple configurations found. Using the first one.")
```
should we create an issue to outline this for a next iteration / help wanted issue on GH?

burtenshaw replied:
Yes that could be useful.

davidberenstein1957 commented on argilla/src/argilla/settings/_io/_hub.py:
```diff
@@ -0,0 +1,203 @@
+                        labels=names,
+                    )
+                )
+                mapping[name].append(name)
```
So, no support for MultiLabel? Might we infer this from the task tag?

burtenshaw replied:
Multilabel is out of scope.

davidberenstein1957 commented on argilla/src/argilla/settings/_io/_hub.py:
```diff
@@ -0,0 +1,203 @@
+                )
+                mapping[name].append(name)
+            if name not in mapped_questions:
+                metadata.append(TermsMetadataProperty(name=f"{name}_metadata", options=names))
```
Wouldn't it be easier to potentially allow for all options? Similar to the approach with Integers and Floats?

burtenshaw replied:
So to leave the `options` param empty? I always thought defining options improved search filters UX, but I might be wrong.

davidberenstein1957 replied:
I am not entirely sure @frascuchon

burtenshaw replied:
The backend supplies unique values if the options are not defined. So I've gone with this.

davidberenstein1957 commented on argilla/src/argilla/settings/_io/_hub.py:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    if not questions:
+        questions.append(TextQuestion(name="comment", required=True))
```
`required=True` by default right?
```suggestion
        questions.append(TextQuestion(name="comment"))
```

burtenshaw replied:
I believe we need at least 1 required question for valid Argilla settings.

burtenshaw committed:
feat: refactor attribute checks to use enums and positive check

burtenshaw committed:
feat: use comprehension in dataset record loader

burtenshaw committed:
test: update test to logic changes in featute mapping

burtenshaw committed:
[FEATURE] from hub with any dataset (#5492)

# Description
<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->

Closes #<issue_number>

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- Refactor (change restructuring the codebase without changing
functionality)
- Improvement (change adding some improvement to an existing
functionality)
- Documentation update

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

burtenshaw committed:
Merge branch 'develop' into feature/features-based-settings

burtenshaw committed:
test: remove retries from integration test

burtenshaw committed:
test: catch FileNotFoundError and retry

burtenshaw committed:
test: reinstate retries in test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
test: increase retry wait

burtenshaw committed:
Merge branch 'feature/features-based-settings' of https://github.com/argilla-io/argilla into feature/features-based-settings

burtenshaw committed:
test: replace test name with function name

burtenshaw committed:
test: drop push to hub from test

burtenshaw committed:
feat: check for image when writing dataset card

burtenshaw committed:
test: use datasets to validate mock dataset

frascuchon commented on argilla/src/argilla/settings/_io/_hub.py:
```diff
@@ -0,0 +1,235 @@
+    url = f"{DATASETS_SERVER_BASE_URL}/info?dataset={repo_id}"
+
+    try:
+        with httpx.Client(timeout=DATASETS_SERVER_TIMEOUT) as client:
```
Just a comment here: 

We can preset some values here to avoid repeat on each call (but we only have one, :_))
```suggestion
        with httpx.Client(base_url= DATASETS_SERVER_BASE_URL, headers=DATASETS_SERVER_HEADERS, timeout=DATASETS_SERVER_TIMEOUT) as client:
```

burtenshaw committed:
feat: add params to httpx client call

Co-authored-by: Paco Aranda <frascuchon@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

burtenshaw committed:
Feature/chat-in-feature-based-settings (#5509)

# Description
<!-- Please include a summary of the changes and the related issue.
Please also include relevant motivation and context. List any
dependencies that are required for this change. -->

```python
import argilla as rg

dataset = rg.Dataset.from_hub(
    repo_id="argilla/magpie-ultra-v0.1",
    name="magpie-ultra",
    split="train[:100]",
)
```

**Type of change**
<!-- Please delete options that are not relevant. Remember to title the
PR according to the type of change -->

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing
functionality to not work as expected)
- Refactor (change restructuring the codebase without changing
functionality)
- Improvement (change adding some improvement to an existing
functionality)
- Documentation update

**How Has This Been Tested**
<!-- Please add some reference about how your feature has been tested.
-->

**Checklist**
<!-- Please go over the list and make sure you've taken everything into
account -->

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature
works
- I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
